File: solutions/739.daily-temperatures.py
#
# @lc app=leetcode id=739 lang=python3
#
# [739] Daily Temperatures
#

# @lc code=start
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        stack = []
        for pos, temp in enumerate(temperatures):
            while stack and temperatures[stack[-1]] < temp:
                prevPos = stack.pop()
                res[prevPos] = pos - prevPos
            stack.append(pos)

        return res

# @lc code=end

File: solutions/739.daily-temperatures.java
/*
 * @lc app=leetcode id=739 lang=java
 *
 * [739] Daily Temperatures
 */

// @lc code=start

import java.util.Stack;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] res = new int[n];
        Stack<Integer> stack = new Stack<>();
        for (int pos = 0; pos < n; pos++) {
            int temp = temperatures[pos];
            while (!stack.isEmpty() && temperatures[stack.peek()] < temp) {
                Integer prevPos = stack.pop();
                res[prevPos] = pos - prevPos;
            }
            stack.push(pos);
        }
        return res;
    }
}
// @lc code=end

